flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImplTest.java

```diff
@@ -0,0 +1,357 @@
+ */
+public class JobAwareShuffleEnvironmentImplTest extends TestLogger {
+
+	private JobID jobId = new JobID();
+	private TestResultPartitionWriter externallyManagedWriter;
+	private TestResultPartitionWriter internallyManagedWriter;
+	private TestShuffleEnvironment wrappedEnvironment;
+	private JobAwareShuffleEnvironmentImpl<?> jobAwareEnvironment;
+	private Collection<ResultPartitionWriter> writers;
+
+	@Before
+	public void setup() {
+		final ExecutionAttemptID executionAttemptId = new ExecutionAttemptID();
+		externallyManagedWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+		internallyManagedWriter = new TestResultPartitionWriter(
+			new ResultPartitionID(new IntermediateResultPartitionID(), executionAttemptId));
+
+		wrappedEnvironment = new TestShuffleEnvironment(
+			Arrays.asList(externallyManagedWriter, internallyManagedWriter));
+
+		jobAwareEnvironment = new JobAwareShuffleEnvironmentImpl<>(wrappedEnvironment);
+
+		ResultPartitionDeploymentDescriptor deploymentDescriptor1 = createResultPartitionDeploymentDescriptor(
+			externallyManagedWriter.resultPartitionID, true);
+		ResultPartitionDeploymentDescriptor deploymentDescriptor2 = createResultPartitionDeploymentDescriptor(
+			internallyManagedWriter.resultPartitionID, false);
+
+		writers = createResultPartitionWriters(
+			jobAwareEnvironment,
+			jobId,
+			executionAttemptId,
+			deploymentDescriptor1,
+			deploymentDescriptor2
+		);
+	}
+
+	@Test
+	public void testInitialState() {
+		final JobID jobId = new JobID();
+		final TestResultPartitionWriter partitionWriter1 = new TestResultPartitionWriter(new ResultPartitionID());
+		final TestResultPartitionWriter partitionWriter2 = new TestResultPartitionWriter(new ResultPartitionID());
+
+		final TestShuffleEnvironment wrappedEnvironment = new TestShuffleEnvironment(
+			Arrays.asList(partitionWriter1, partitionWriter2));
+
+		final JobAwareShuffleEnvironmentImpl jobAwareEnvironment = new JobAwareShuffleEnvironmentImpl(wrappedEnvironment);
+
+		assertFalse(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+	}
+
+	@Test
+	public void testReleaseAllPartitions() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
+
+		assertTrue(jobAwareEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+		jobAwareEnvironment.releaseAllFinishedPartitionsForJobAndMarkJobInactive(jobId);
+		assertThat(wrappedEnvironment.releasedPartitions, contains(externallyManagedWriter.resultPartitionID));
+	}
+
+	@Test
+	public void testReleasePartitions() throws IOException {
+		jobAwareEnvironment.markJobActive(jobId);
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.setup();
+		}
+
+		for (ResultPartitionWriter writer : writers) {
+			writer.finish();
+		}
```

Comment:
The above codes for writer setup and finish might be extracted in separate method for reusing in tests.